rfcs/0064-autotvm-tune-with-subgraph-granularity.md

```diff
@@ -0,0 +1,135 @@
+
+```python
+def select_implementation():
+  # ignore some codes
+
+  if GLOBAL_SCOPE.tune_subgraph:
+    # In some cases, one strategy's compute may call another compute.
+    # So the impl name need to match with actual compute.
+    if workloads[best_plevel_impl]:
+        workload = workloads[best_plevel_impl]
+        if best_plevel_impl.name != "injective.cpu" and best_plevel_impl.name != workload[0]:
+          best_plevel_impl.name = workload[0]
+  # value changed in python side will not effect C++ side,
+  # so here need to pass new name to C++
+  return best_plevel_impl, outputs[best_plevel_impl], best_plevel_impl.name
+```
+
+Because the value changed on the python side will not affect the C++ side, we need to pass the new name to C++. This results in the function `select_implementation` number of returned values changing to three. This should be noticed.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are three cases in which tuning with subgraph may not get better performance.
+
+Case 1: Anchor operator has more than one implementation.
+We register the subgraph tuning task by `outputs` in step 1.4. No matter how many implementations the anchor operator has, step 1.1 will only pick the implementation with the highest level and return outputs computed by it. So the subgraph tuning tasks may not contain the potential best implementation.
+
+Case 2: Anchor operator's function `fcompute` needs value from config such as code block below. In step 2.2, computing output will call function `_pack_data` and the `cfg` suppose to be the best config of the subgraph. But in step 2.2 we don't know which subgraph the anchor operator belongs to yet, so we cannot get the right config from the best history and fallback to the default one. This may bring great performance regression.
+
+```python
+def _pack_data(cfg, data, kernel):
+    n, _, ih, iw = get_const_tuple(data.shape)
+    oc, ic, kh, kw = get_const_tuple(kernel.shape)
+    ic_bn, oc_bn = cfg["tile_ic"].size[-1], cfg["tile_oc"].size[-1]
+    ......
+```
+
+Case 3: During task extraction subgraph is lowered by VMCompile, and pass `AlterOpLayout` is disabled, see [code](https://github.com/apache/tvm/blob/main/python/tvm/autotvm/task/relay_integration.py#:~:text=with%20tvm.transform.PassContext(opt_level%3Dopt_level%2C%20disabled_pass%3D%7B%22AlterOpLayout%22%7D)%3A). So during building phash we need to disable pass `AlterOpLayout` too, otherwise the subgraphs generated in the task extracting phase may be different from those generated in the building phase.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The core problem of tune subgraphs is getting output tensors of subgraphs. Employing function `LowerToTECompute::Lower` can achieve minimal change to the current framework.
+
+# Prior art
+[prior-art]: #prior-art
+
+Our implementation is inspired by auto-schedule.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
```

Comment:
Same question here. This also determines how general the tuned schedule could be. For example, considering the following hashing approaches:

H1: Similar to auto-scheduler, such as `["impl-name", "args", "compute DAG MD5"]`. This guarantees that the tuned schedule will be applied to ONLY exactly the same subgraph.

H2 Canonicalize compute DAG. For example, canonicalize a compute DAG `cast - dense - cast` to be `elementwise - dense - elementwise`. Since different elementwise ops in this DAG should result in almost the same performance, canonicalization makes the tuned result more general applicable. The drawback is we need to be careful to make sure the canonicalized compute DAG won't cover too many undesired cases.

Reply:
> H1: Similar to auto-scheduler, such as `["impl-name", "args", "compute DAG MD5"]`. This guarantees that the tuned schedule will be applied to ONLY exactly the same subgraph.

We use the first approach. We serialize the `iotensors` of the subgraph to a hashable tuple and compute the tuple MD5 for getting a hash key. Then use `subgraph name + hash key` as the subgraph task name.